openedx/core/lib/block_cache/block_structure.py

```diff
@@ -0,0 +1,591 @@
+"""
+Module with family of classes for block structures.
+    BlockStructure - responsible for block existence and relations.
+    BlockStructureBlockData - responsible for block & transformer data.
+    BlockStructureXBlockData - responsible for xBlock data.
+
+The following internal data structures are implemented:
+    _BlockRelations - Data structure for a single block's relations.
+    _BlockData - Data structure for a single block's data.
+"""
+from collections import defaultdict
+from logging import getLogger
+
+from openedx.core.lib.graph_traversals import traverse_topologically, traverse_post_order
+
+from .exceptions import TransformerException
+
+
+logger = getLogger(__name__)  # pylint: disable=C0103
+
+
+# A dictionary key value for storing a transformer's version number.
+TRANSFORMER_VERSION_KEY = '_version'
+
+
+class _BlockRelations(object):
+    """
+    Data structure to encapsulate parents and children relationship for
+    a single block.
+    """
+    def __init__(self):
+
+        # List of usage keys of this block's parents.
+        # list [UsageKey]
+        self.parents = []
+
+        # List of usage keys of this block's children.
+        # list [UsageKey]
+        self.children = []
+
+
+class BlockStructure(object):
+    """
+    Base class for a block structure.  BlockStructures are constructed
+    using the BlockStructureFactory and then used as the currency across
+    Transformers.
+
+    This base class keeps track of the block structure's root_block_key,
+    the existence of the blocks, and their parents and children
+    relationships (graph nodes and edges).
+    """
+    def __init__(self, root_block_key):
+
+        # The usage key of the root block for this structure.
+        # UsageKey
+        self.root_block_key = root_block_key
+
+        # Map of a block's usage key to its block relations. The
+        # existence of a block in the structure is determined by its
+        # presence in this map.
+        # defaultdict {UsageKey: _BlockRelations}
+        self._block_relations = defaultdict(_BlockRelations)
+
+        # Add the root block.
+        self._add_block(self._block_relations, root_block_key)
+
+    def __iter__(self):
+        """
+        The default iterator for a block structure is a topological
```

Comment:
To check understanding, the default is topological because 1. it's the common case. 2. If you don't understand the difference it's a safer default?

I ask because I'm generally in favor of asking people to choose explicitly for things like this if there's not an obvious choice just from knowing the data structure.